### PR TITLE
Expose Flush, Update docs & Readme

### PR DESCRIPTION
**Pull request by mangalaman93**

### CODE REVIEW

**Comment by mangalaman93** [comment]
@rohansuri please review once.

File: README.md [hunk 1]
```diff
@@ -40,17 +40,21 @@ defer bq.Close()
 In this case, bigqueue will never allocate more memory than `4KB*10=40KB`. This
 memory is above and beyond the memory used in buffers for copying data.
 
-Bigqueue allows to set flush intervals based on either elapsed time or number of mutate operations.
-Flush syncs the in memory changes of memory mapped files with disk.  
-Note: This is a best effort flush and elapsed time and number of mutate operations are checked upon an enqueue/dequeue.
+Bigqueue allows to set periodic flush intervals based on either elapsed time or
+number of mutate (enqueue/deqeueu) operations. Flush syncs the in memory changes
```
Comment (rohansuri) [inline README.md]: deqeueu typo

File: bigqueue.go [hunk 2]
```diff
@@ -106,16 +106,23 @@ func (q *MmapQueue) Close() error {
...
-	shouldFlush := q.mutateOpsSinceFlush >= q.conf.flushIntervalMutateOps
-	shouldFlush = shouldFlush || clock.Since(q.prevFlushTime) >= q.conf.flushElapsedDuration
-	if shouldFlush {
-		if err := q.am.flush(); err != nil {
-			return err
-		}
-		q.mutateOpsSinceFlush = 0
-		q.prevFlushTime = clock.Now()
+// Flush syncs the in memory content of bigqueue to disk
+func (q *MmapQueue) Flush() error {
```
Comment (rohansuri) [inline bigqueue.go]: should we add a call to Flush as part of this test [TestResetOfBothFlushIntervalStates](https://github.com/grandecola/bigqueue/blob/master/bigqueue_flush_test.go#L151)?
just to assert that either of: enoughTime, enoughMutate, or explicit flush results in a reset of the flush interval state.

**Comment by rohansuri** [comment]
just two tiny comments.

**Comment by mangalaman93** [comment]
@rohansuri updated the PR. Good idea to add a test, realized that the interface didn't have Flush function in it. Also, updated variables to have shorted names, check now.

File: bigqueue_flush_test.go [hunk 3]
```diff
@@ -14,97 +14,91 @@ import (
...
 
-func TestSetFlushElapsedDurationValidation(t *testing.T) {
-	testDir := path.Join(os.TempDir(), fmt.Sprintf("testdir_%d", rand.Intn(1000)))
+func TestSetPeriodicFlushDurationFail(t *testing.T) {
+	testDir := path.Join(os.TempDir(), fmt.Sprintf("testdir_%v", rand.Intn(1000)))
 	createTestDir(t, testDir)
 	defer deleteTestDir(t, testDir)
 
-	_, err := NewMmapQueue(testDir, SetFlushElapsedDuration(-1))
+	_, err := NewMmapQueue(testDir, SetPeriodicFlushDuration(-1))
```
Comment (rohansuri) [inline bigqueue_flush_test.go]: just realized I should've made `SetPeriodicFlushDuration(0)` here. to test against the highest non-allowed value.

File: bigqueue_flush_test.go [hunk 4]
```diff
@@ -169,37 +160,42 @@ func TestResetOfBothFlushIntervalStates(t *testing.T) {
...
+	checkLastFlushTime(bq, t, clock.Now())
 
-	// flush after 2 mutateOps, causes mutateOps reset and time reset to Now()
-
-	checkMutateOps(bq, t, 0)
-	prevFlushTime = (bq.(*MmapQueue)).prevFlushTime
-	if prevFlushTime != clock.Now() {
-		t.Fatalf("expected prevFlushTime: %v to be reset to clock's now: %v", prevFlushTime, clock.Now())
+	// move time ahead and mutate some ops
+	clock.Advance(flushPeriod)
```
Comment (rohansuri) [inline bigqueue_flush_test.go]: remove this advance and check mutops to be 1 after the enqueue.
just to show better than mutops went from 0 -> 1 -> 0.

File: bigqueue_flush_test.go [hunk 5]
```diff
@@ -169,37 +160,42 @@ func TestResetOfBothFlushIntervalStates(t *testing.T) {
...
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
 	}
+	checkMutOps(bq, t, 0)
+	checkLastFlushTime(bq, t, clock.Now())
 
+	clock.Advance(flushPeriod)
```
Comment (rohansuri) [inline bigqueue_flush_test.go]: advance only by flushPeriod/2 or something less?
to assert that bq.Flush() is an unconditional flush.

**Comment by rohansuri** [comment]
just two tiny test assertion suggestions.
everything else is neat.

**Comment by mangalaman93** [comment]
@rohansuri made the changes

**Comment by rohansuri** [comment]
Yup looking good to merge now.

**Pull request merged**